oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/mongo/MongoClusterListener.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.jackrabbit.oak.plugins.document.mongo;
+
+import com.mongodb.ServerAddress;
+import com.mongodb.connection.ServerConnectionState;
+import com.mongodb.connection.ServerDescription;
+import com.mongodb.event.ClusterDescriptionChangedEvent;
+import com.mongodb.event.ClusterListener;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+public class MongoClusterListener implements ClusterListener {
+
+    // Sometimes we need to wait a few seconds in case the connection was just created, the listener
+    // didn't have time to receive the description from the cluster. This latch is used to check
+    // if the connection was properly initialized.
+    private final CountDownLatch LATCH;
+    private final int LATCH_AWAIT_TIMEOUT = 15;
+    private boolean replicaSet = false;
+    private ServerAddress serverAddress;
+    private ServerAddress primaryAddress;
+
+    public MongoClusterListener() {
+        LATCH = new CountDownLatch(1);
+    }
+
+    @Nullable
+    public ServerAddress getServerAddress() {
+        waitForLatch();
+        return serverAddress;
+    }
+
+    @Nullable
+    public ServerAddress getPrimaryAddress() {
+        waitForLatch();
+        return primaryAddress;
+    }
+
+    public boolean isReplicaSet() {
+        waitForLatch();
+        return replicaSet;
+    }
+
+    @Override
+    public void clusterDescriptionChanged(final ClusterDescriptionChangedEvent event) {
+        for (ServerDescription sd : event.getNewDescription().getServerDescriptions()) {
```

Comment:
```suggestion
        Optional<ServerDescription> ol = Iterables.tryFind(event.getNewDescription().getServerDescriptions(), s -> s.getState() == ServerConnectionState.CONNECTED);
        if (ol.isPresent()) {
            ServerDescription sd = ol.get();
            serverAddress = sd.getAddress();
            primaryAddress = new ServerAddress(sd.getPrimary());
            if (sd.isReplicaSetMember()) {
                // Can't assign directly the result of the function because in some cases the cluster
                // type is UNKNOWN, mainly when the cluster is changing it's PRIMARY.
                replicaSet = true;
            }
            latch.countDown();
        } else {
            // log here that no connected state is present
        }
```

Reply:
I have applied your recommendation. However, I don't think we should add a log message when no connected state is present, as this event can be executed in multiple situations (for example, when there is a disconnection), meaning we could flood the logs with messages that are not useful.

What I'm adding now is a single log message when `getPrimaryAddress()`, `getAddress()` or `isReplicaSet()` methods return a null value because the connection was not properly established. This way we ensure the log message will show up just once.